### PR TITLE
Raster labeling

### LINKED ISSUE #14408: Show raster values on screen when zoomed-in enough.
Author Name: **Thaddeus -** (Thaddeus -)
Original Redmine Issue: [4481](https://issues.qgis.org/issues/4481)

Redmine category:rasters

---
Please see attached images.

---

- [S_1.png](https://issues.qgis.org/attachments/download/3522/S_1.png) (Thaddeus -)
- [S_2.png](https://issues.qgis.org/attachments/download/3524/S_2.png) (Thaddeus -)
- [S_3.png](https://issues.qgis.org/attachments/download/3525/S_3.png) (Thaddeus -)
- [S_4.png](https://issues.qgis.org/attachments/download/3526/S_4.png) (Thaddeus -)

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows raster pixels to be labeled with the value taken from a raster band.

Labels are registered with the labeling engine, so participate in label conflict resolution and overlap avoidance

Options include

- Selection of band to take values from
- Using QgsNumericFormat to customise the number format for the labels
- Uses text renderer, so supports buffers, shadows, etc
- Label priority
- Scale dependant visibility
- Optional pixel size dependent visibilty, ie show only when pixels are > 4mm in size
- Z index control, overlap avoidance mode
- Option for showing values resampled over neighbouring pixels

Fixes https://github.com/qgis/QGIS/issues/14408

Funded by Canton de Neuchâtel

https://github.com/user-attachments/assets/1f1cdaff-9ace-483b-b204-33069d967eac

### CODE REVIEW

**Comment by saberraz** [comment]
Very nice feature.  A couple of silly questions maybe:
- Does this work with expression engine. Lets say I have a temporal raster where times are stored as a band. Can I then show the raster pixel based on the band and then see the label changes as I move the temporal controller?!
- Will there be an option to merge the same labels for neighboring cells?

**Comment by nyalldawson** [comment]
@saberraz 
> Does this work with expression engine. Lets say I have a temporal raster where times are stored as a band. Can I then show the raster pixel based on the band and then see the label changes as I move the temporal controller?!

No -- right now the band selection is fixed, and you get the pixel values only. (There's kind-of expression support via the "custom expression" option in the number format panel, but that's limited).

I think it'd be nice for a follow up to add an expression based labels mode, so eg you can get all band values as a list and choose how to display these (but that's not in my current scope).

> Will there be an option to merge the same labels for neighboring cells?

It's not planned, but I do have a follow up coming which will allow the labels to show downsampled values (eg labeling only every 2 second pixel)

**Comment by saberraz** [comment]
Great work and thanks for the explanation.

**Comment by nyalldawson** [comment]
I would LOVE a review before I leave on Christmas holidays on Friday please! (Otherwise this will likely miss 3.42 deadline, I'm away from my computer for a month)

**Comment by nirvn** [comment]
@nyalldawson , I'll do it.

**Review by nirvn — COMMENTED** [review]
A couple of comments.

We'll need the labeling panel to also be featured in the raster properties dialog before we ship the functionality.

But beyond that, big thanks! It's something I was looking for multiple times.

File: src/core/raster/qgsrasterlabeling.cpp [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+
+QgsRasterLayerLabelProvider::QgsRasterLayerLabelProvider( QgsRasterLayer *layer )
+  : QgsAbstractLabelProvider( layer )
+  , mNumericFormat( std::make_unique< QgsBasicNumericFormat >() )
+{
+  mPlacement = Qgis::LabelPlacement::OverPoint;
+  mFlags |= DrawLabels;
+
+  // always consider these labels highest priority
+  // TODO possibly expose as a setting
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/raster/qgsrasterlabeling.cpp:42]: Oh, that's already done -- it's an outdated comment!

File: src/core/raster/qgsrasterlabeling.cpp [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+    thinningElem.setAttribute( QStringLiteral( "limitNumLabels" ), mThinningSettings.limitNumberOfLabelsEnabled() );
+    thinningElem.setAttribute( QStringLiteral( "minFeatureSize" ), mThinningSettings.minimumFeatureSize() );
+    elem.appendChild( thinningElem );
+  }
+
+  {
+    QDomElement renderingElem = doc.createElement( QStringLiteral( "rendering" ) );
+    renderingElem.setAttribute( QStringLiteral( "scaleVisibility" ), mScaleVisibility );
+    renderingElem.setAttribute( QStringLiteral( "scaleMin" ), mMaximumScale );
+    renderingElem.setAttribute( QStringLiteral( "scaleMax" ), mMinimumScale );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/raster/qgsrasterlabeling.cpp:453]: yeah, it's come from the vector labeling. I'm not sure if we'll ever want to be able to copy/paste label settings between vector/raster, but if so we'd need it.

File: src/gui/raster/qgsrasterlabelingwidget.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    //! Sets the layer to configure
+    void setLayer( QgsMapLayer *layer );
+
+    /**
+     * Saves the labeling configuration to the destination layer.
+     */
+    void writeSettingsToLayer();
+
+    //! Saves the labeling configuration and immediately updates the map canvas to reflect the changes
+    void apply() override;
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/gui/raster/qgsrasterlabelingwidget.h:60]: Yeah, I'd like to defer that until we've a QEP for code style in place + we have a bulk tool to fix it across the whole codebase.

File: src/gui/raster/qgsrasterlabelsettingswidget.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+  const int prevIndex = mOptionsTab->currentIndex();
+
+  setPropertyOverrideButtonsVisible( true );
+  mTextFormatsListWidget->setEntityTypes( QList<QgsStyle::StyleEntity>() << QgsStyle::TextFormatEntity );
+
+  delete mLabelingOptionsListWidget->takeItem( 6 ); // callouts
+  delete mLabelingOptionsListWidget->takeItem( 3 ); // mask
+
+  mOptionsTab->removeTab( 6 );
+  mOptionsTab->removeTab( 3 );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/gui/raster/qgsrasterlabelsettingswidget.cpp:89]: Ok, addressed

**Comment by nyalldawson** [comment]
@nirvn 
> We'll need the labeling panel to also be featured in the raster properties dialog before we ship the functionality.

Ok, done!

File: src/app/qgslayerstylingwidget.h [hunk 5]
```diff
@@ -171,6 +172,7 @@ class APP_EXPORT QgsLayerStylingWidget : public QWidget, private Ui::QgsLayerSty
     QgsMapLayer *mCurrentLayer = nullptr;
     QgsLabelingWidget *mLabelingWidget = nullptr;
     QgsMeshLabelingWidget *mMeshLabelingWidget = nullptr;
+    QPointer<QgsRasterLabelingWidget> mRasterLabelingWidget;
```
Comment (m-kuhn) [inline src/app/qgslayerstylingwidget.h:175]: Any reason to make this one a `QPointer` while the others are not?

Comment (nyalldawson) [reply src/app/qgslayerstylingwidget.h:175]: It seems cleaner to me. The others should probably be too

**Pull request merged**